### PR TITLE
Min max size functions added

### PR DESCRIPTION
**Pull request by DaddyWesker**
So, I've added three functions: size-atom, max-atom and min-atom. Works as supposed by the name:

!(size-atom (5 4 3))
[3]
!(max-atom (4 3 4.4))
[4.4]
!(min-atom (3 5 2.9))
[2.9]

Documentation and tests are also added.

### CODE REVIEW

File: lib/src/metta/runner/stdlib.rs [hunk 1]
```diff
@@ -1161,6 +1161,94 @@ impl CustomExecute for IntersectionAtomOp {
...
+        for x in children.iter() {
+            match AsPrimitive::from_atom(x).as_number() {
+                None => Err(ExecError::from("Only numbers allowed in expression")),
+                _ => Ok({}),
+            }?
+        };
+        match children.into_iter().map(|x| Into::<f64>::into(AsPrimitive::from_atom(x).as_number().unwrap())).reduce(f64::min) {
+            Some(min) => Ok(vec![Atom::gnd(Number::Float(min))]),
+            None      => Err(ExecError::from("Empty expression")),
+        }
```
Comment (vsbogd) [inline lib/src/metta/runner/stdlib.rs:1224]: 1. One doesn't need to iterate through children twice, `fold` allows doing this in one step
2. `Into::<>` is needed only when compiler cannot inference the proper type from context, here it is not needed
I would suggest replacing code by the following:
```suggestion
        if children.is_empty() {
            Err(ExecError::from("Empty expression"))
        } else {
            children.into_iter().fold(Ok(f64::INFINITY), |res, x| {
                match (res, AsPrimitive::from_atom(x).as_number()) {
                    (res @ Err(_), _) => res,
                    (_, None) => Err(ExecError::from("Only numbers are allowed in expression")),
                    (Ok(min), Some(x)) => Ok(f64::min(min, x.into())),
                }
            })
        }.map(|min| vec![Atom::gnd(Number::Float(min))])
```

File: lib/src/metta/runner/stdlib.rs [hunk 2]
```diff
@@ -1161,6 +1161,94 @@ impl CustomExecute for IntersectionAtomOp {
...
+        for x in children.iter() {
+            match AsPrimitive::from_atom(x).as_number() {
+                None => Err(ExecError::from("Only numbers allowed in expression")),
+                _ => Ok({}),
+            }?
+        };
+        match children.into_iter().map(|x| Into::<f64>::into(AsPrimitive::from_atom(x).as_number().unwrap())).reduce(f64::max) {
+            Some(max) => Ok(vec![Atom::gnd(Number::Float(max))]),
+            None      => Err(ExecError::from("Empty expression")),
+        }
```
Comment (vsbogd) [inline lib/src/metta/runner/stdlib.rs:1192]: See `min-atom` comments:
```suggestion
        if children.is_empty() {
            Err(ExecError::from("Empty expression"))
        } else {
            children.into_iter().fold(Ok(f64::NEG_INFINITY), |res, x| {
                match (res, AsPrimitive::from_atom(x).as_number()) {
                    (res @ Err(_), _) => res,
                    (_, None) => Err(ExecError::from("Only numbers are allowed in expression")),
                    (Ok(max), Some(x)) => Ok(f64::max(max, x.into())),
                }
            })
        }.map(|max| vec![Atom::gnd(Number::Float(max))])
```

File: lib/src/metta/runner/stdlib.rs [hunk 3]
```diff
@@ -1161,6 +1161,94 @@ impl CustomExecute for IntersectionAtomOp {
...
+}
+
+#[derive(Clone, Debug)]
+pub struct MinAtomOp {}
+
+grounded_op!(MinAtomOp, "min-atom");
+
+impl Grounded for MinAtomOp {
+    fn type_(&self) -> Atom {
+        Atom::expr([ARROW_SYMBOL, ATOM_TYPE_EXPRESSION, ATOM_TYPE_ATOM])
```
Comment (vsbogd) [inline lib/src/metta/runner/stdlib.rs:1203]:
```suggestion
        Atom::expr([ARROW_SYMBOL, ATOM_TYPE_EXPRESSION, ATOM_TYPE_NUMBER])
```

File: lib/src/metta/runner/stdlib.rs [hunk 4]
```diff
@@ -1161,6 +1161,94 @@ impl CustomExecute for IntersectionAtomOp {
...
 }
 
+#[derive(Clone, Debug)]
+pub struct MaxAtomOp {}
+
+grounded_op!(MaxAtomOp, "max-atom");
+
+impl Grounded for MaxAtomOp {
+    fn type_(&self) -> Atom {
+        Atom::expr([ARROW_SYMBOL, ATOM_TYPE_EXPRESSION, ATOM_TYPE_ATOM])
```
Comment (vsbogd) [inline lib/src/metta/runner/stdlib.rs:1171]:
```suggestion
        Atom::expr([ARROW_SYMBOL, ATOM_TYPE_EXPRESSION, ATOM_TYPE_NUMBER])
```

File: lib/src/metta/runner/stdlib.rs [hunk 5]
```diff
@@ -2122,6 +2216,34 @@ mod tests {
...
+        let res = SizeAtomOp{}.execute(&mut vec![expr!()]).expect("No result returned");
+        assert_eq!(res, vec![expr!({Number::Integer(0)})]);
+    }
+
+    #[test]
+    fn min_atom_op() {
+        let res = MinAtomOp{}.execute(&mut vec![expr!({Number::Integer(5)} {Number::Integer(4)} {Number::Float(5.5)})]).expect("No result returned");
+        assert_eq!(res, vec![expr!({Number::Integer(4)})]);
+        let res = MinAtomOp{}.execute(&mut vec![expr!({Number::Integer(5)} {Number::Integer(4)} "A")]);
+        assert_eq!(res, Err(ExecError::from("Only numbers allowed in expression")));
```
Comment (vsbogd) [inline lib/src/metta/runner/stdlib.rs:2232]:
```suggestion
        assert_eq!(res, Err(ExecError::from("Only numbers are allowed in expression")));
```

File: lib/src/metta/runner/stdlib.rs [hunk 6]
```diff
@@ -2122,6 +2216,34 @@ mod tests {
...
+        let res = MinAtomOp{}.execute(&mut vec![expr!()]);
+        assert_eq!(res, Err(ExecError::from("Empty expression")));
+    }
+
+    #[test]
+    fn max_atom_op() {
+        let res = MaxAtomOp{}.execute(&mut vec![expr!({Number::Integer(5)} {Number::Integer(4)} {Number::Float(5.5)})]).expect("No result returned");
+        assert_eq!(res, vec![expr!({Number::Float(5.5)})]);
+        let res = MaxAtomOp{}.execute(&mut vec![expr!({Number::Integer(5)} {Number::Integer(4)} "A")]);
+        assert_eq!(res, Err(ExecError::from("Only numbers allowed in expression")));
```
Comment (vsbogd) [inline lib/src/metta/runner/stdlib.rs:2242]:
```suggestion
        assert_eq!(res, Err(ExecError::from("Only numbers are allowed in expression")));
```

File: lib/src/metta/runner/stdlib_minimal.rs [hunk 7]
```diff
@@ -621,6 +627,26 @@ mod tests {
...
+        assert_eq!(run_program(&format!("!(min-atom (5 4 5.5))")), Ok(vec![vec![expr!({Number::Integer(4)})]]));
+        assert_eq!(run_program(&format!("!(min-atom ())")), Ok(vec![vec![expr!("Error" ({ stdlib::MinAtomOp{} } ()) "Empty expression")]]));
+        assert_eq!(run_program(&format!("!(min-atom (3 A B 5))")), Ok(vec![vec![expr!("Error" ({ stdlib::MinAtomOp{} } ({Number::Integer(3)} "A" "B" {Number::Integer(5)})) "Only numbers allowed in expression")]]));
+    }
+
+    #[test]
+    fn metta_max_atom() {
+        assert_eq!(run_program(&format!("!(max-atom (5 4 5.5))")), Ok(vec![vec![expr!({Number::Float(5.5)})]]));
+        assert_eq!(run_program(&format!("!(max-atom ())")), Ok(vec![vec![expr!("Error" ({ stdlib::MaxAtomOp{} } ()) "Empty expression")]]));
+        assert_eq!(run_program(&format!("!(max-atom (3 A B 5))")), Ok(vec![vec![expr!("Error" ({ stdlib::MaxAtomOp{} } ({Number::Integer(3)} "A" "B" {Number::Integer(5)})) "Only numbers allowed in expression")]]));
```
Comment (vsbogd) [inline lib/src/metta/runner/stdlib_minimal.rs:647]:
```suggestion
        assert_eq!(run_program(&format!("!(max-atom (3 A B 5))")), Ok(vec![vec![expr!("Error" ({ stdlib::MaxAtomOp{} } ({Number::Integer(3)} "A" "B" {Number::Integer(5)})) "Only numbers are allowed in expression")]]));
```

File: lib/src/metta/runner/stdlib_minimal.rs [hunk 8]
```diff
@@ -621,6 +627,26 @@ mod tests {
...
+    fn metta_size_atom() {
+        assert_eq!(run_program(&format!("!(size-atom (5 4 3 2 1))")), Ok(vec![vec![expr!({Number::Integer(5)})]]));
+        assert_eq!(run_program(&format!("!(size-atom ())")), Ok(vec![vec![expr!({Number::Integer(0)})]]));
+    }
+
+    #[test]
+    fn metta_min_atom() {
+        assert_eq!(run_program(&format!("!(min-atom (5 4 5.5))")), Ok(vec![vec![expr!({Number::Integer(4)})]]));
+        assert_eq!(run_program(&format!("!(min-atom ())")), Ok(vec![vec![expr!("Error" ({ stdlib::MinAtomOp{} } ()) "Empty expression")]]));
+        assert_eq!(run_program(&format!("!(min-atom (3 A B 5))")), Ok(vec![vec![expr!("Error" ({ stdlib::MinAtomOp{} } ({Number::Integer(3)} "A" "B" {Number::Integer(5)})) "Only numbers allowed in expression")]]));
```
Comment (vsbogd) [inline lib/src/metta/runner/stdlib_minimal.rs:640]:
```suggestion
        assert_eq!(run_program(&format!("!(min-atom (3 A B 5))")), Ok(vec![vec![expr!("Error" ({ stdlib::MinAtomOp{} } ({Number::Integer(3)} "A" "B" {Number::Integer(5)})) "Only numbers are allowed in expression")]]));
```

**Comment by vsbogd** [comment]
Overall looks good to me, added some suggestions for documentation and code change to eliminate double iteration through expression.

**Comment by DaddyWesker** [comment]
> Overall looks good to me, added some suggestions for documentation and code change to eliminate double iteration through expression.

Thanks for the code changes, now it looks much better. I've pushed those changes.

**Review by vsbogd — APPROVED** [review]

**Pull request merged**